com.ibm.streamsx.topology/com.ibm.streamsx.topology.functional.python/PyFunctionMultiTransform/PyFunctionMultiTransform_cpp.cgt

```diff
@@ -0,0 +1,187 @@
+
+  streamsx::topology::Splpy::loadCPython(spl_setup_py);
+
+  PyGILState_STATE gstate;
+  gstate = PyGILState_Ensure();
+
+    PyObject *_module_;
+    PyObject *_function_;
+
+    std::string appDirSetup = "import streamsx.topology.runtime\n";
+    appDirSetup += "streamsx.topology.runtime.setupOperator(\"";
+    appDirSetup += ProcessingElement::pe().getApplicationDirectory();
+    appDirSetup += "\")\n";
+
+    const char* spl_setup_appdir = appDirSetup.c_str();
+    if (PyRun_SimpleString(spl_setup_appdir) != 0) {
+         SPLAPPTRC(L_ERROR, "Python script splpy_setup.py failed!", "python");
+         PyErr_Print();
+         throw;
+    }
+
+<%
+ my $module =  $model->getParameterByName("functionModule")->getValueAt(0)->getCppExpression() . '.c_str()';
+ my $functionName = $model->getParameterByName("functionName")->getValueAt(0)->getCppExpression() . '.c_str()';
+%>
+
+    PyObject * appFunction = 
+      streamsx::topology::Splpy::loadFunction(<%=$module%>, <%=$functionName%>);
+
+<% if ($pystyle eq 'python') { %>
+
+    PyObject * depickleInput = 
+      streamsx::topology::Splpy::loadFunction("streamsx.topology.runtime", "depickleInput");
+
+    PyObject * funcArg = PyTuple_New(1);
+    PyTuple_SetItem(funcArg, 0, appFunction);
+
+    function_ = PyObject_CallObject(depickleInput, funcArg);
+    // TODO error check
+    Py_DECREF(depickleInput);
+    Py_DECREF(funcArg);
+    
+    iterableObjectFunction_ = 
+      streamsx::topology::Splpy::loadFunction("streamsx.topology.runtime", "iterableObject");
+
+<%}%>
+<% if ($pystyle eq 'string') { %>
+    function_ = appFunction;
+<%}%>
+
+    PyGILState_Release(gstate);
+}
+
+// Destructor
+MY_OPERATOR::~MY_OPERATOR() 
+{
+    // Finalization code goes here
+    if (function_) {
+      Py_DECREF(function_);
+    }
+    if (iterableObjectFunction_) {
+      Py_DECREF(iterableObjectFunction_);
+    }
+}
+
+// Notify port readiness
+void MY_OPERATOR::allPortsReady() 
+{
+}
+ 
+// Notify pending shutdown
+void MY_OPERATOR::prepareToShutdown() 
+{
+    // This is an asynchronous call
+}
+
+// Processing for source and threaded operators   
+void MY_OPERATOR::process(uint32_t idx)
+{
+}
+
+// Tuple processing for mutating ports 
+void MY_OPERATOR::process(Tuple & tuple, uint32_t port)
+{
+}
+
+// Tuple processing for non-mutating ports
+void MY_OPERATOR::process(Tuple const & tuple, uint32_t port)
+{
+  IPort0Type const &ip = static_cast<IPort0Type const &>(tuple);
+
+<% if ($pystyle eq 'python') { %>
+  SPL::blob value = ip.get___spl_po();
+<%}%>
+
+<% if ($pystyle eq 'string') { %>
+  SPL::rstring value = ip.get_string();
+<%}%>
+
+  std::auto_ptr<SPL::blob> pBlob;
+  PyGILState_STATE gstate;
+  gstate = PyGILState_Ensure();
+
+  // invoke python application function and get return value
+  PyObject * pyBytes  = streamsx::topology::Splpy::pyBlobToBytes(value);
+  PyObject * pyReturnVar = streamsx::topology::Splpy::pyTupleFunc(function_, pyBytes);

```

Comment:
i think a follow on change could simplify this into a single function call into Python. That's what the other operators do so it would be good to follow that pattern. (see comment about two functions).

That is the call to get the iterator from the input tuple and to return the function that returns the pickled values can all be done in Python.